res/controllers/Denon-MC4000-scripts.js

```diff
@@ -0,0 +1,544 @@
+}
+
+MC4000.OnVuMeterChangeR = function(value, group, control) {
+    midi.sendShortMsg(0xBF, 0x45, value*0x7F);
+}
+
+MC4000.Deck = function (channel) {
+    // Some state variables
+    this.isVinylMode = false;
+    this.rateMsbValue = 0;
+    this.midiChannel = channel;
+    this.engineChannel = channel + 1;
+    this.group = '[Channel' + this.engineChannel + ']';
+    this.currentRateIndex = 0;
+    
+    // Set default rate range (to +/- 8% like Serato and Technics)
+    engine.setValue(this.group, "rateRange", MC4000.rateRanges[this.currentRateIndex]);
+    // Initialize vinyl mode LED
+    midi.sendShortMsg(0x90 + channel, 0x07, this.isVinylMode ? 0x7F: 0x00);
+    // Callback for touching the jog wheel platter
+    this.platterTouch = function (channel, control, value, status, group) {
+        if (this.isVinylMode) {
+            if (value > 0) {
+                engine.scratchEnable(this.engineChannel, MC4000.jogWheelTicksPerRevolution, 
+                    MC4000.scratchParams.recordSpeed , MC4000.scratchParams.alpha, MC4000.scratchParams.beta);
+            } else { 
+                engine.scratchDisable(this.engineChannel);
+            }
+        }
+    }
+    // Callback for toggling the vinyl mode button
+    this.vinylModeToggle = function (channel, control, value, status, group) {
+        if (value > 0) {
+            this.isVinylMode = !this.isVinylMode;
+            midi.sendShortMsg(0x90 + channel, 0x07, this.isVinylMode ? 0x7F: 0x00);
+        }
+    }
+    // Key lock toggle button
+    this.keylock = new ToggleButton([0x90 + channel, 0x0D], this.group, 'keylock');
+    // Rate range toggle button callback
+    this.rateRange = function(midichan, control, value, status, group) {
+        if (value == 0) {
+            return; // don't respond to note off messages
+        }
+        this.currentRateIndex++;
+        if (this.currentRateIndex >= MC4000.rateRanges.length) {
+            this.currentRateIndex = 0;
+        }
+        engine.setValue(this.group, "rateRange", MC4000.rateRanges[this.currentRateIndex]);
+    };
+    // Sync toggle button
+    // TODO: shift + sync should be used to diable sync instead of using toggle
+    this.sync = new ToggleButton([0x90 + channel, 0x02], this.group, 'sync_enabled');
+    // Cue (transport) button
+    this.cue = new CueButton([0x90 + channel, 0x01], this.group);
+    this.goToStart = new Control([0x90 + channel, 0x05], this.group, // cue shifted
+                            ['start_stop', function () { return 1; }],
+                            ['play_indicator', function (val) { return val * 127; } ]);
+    // Play button
+    this.play = new PlayButton([0x90 + channel, 0x00], this.group);
+    // Headphone cue (PFL) button
+    this.pfl = new ToggleButton([0x90 + channel, 0x1B], this.group, 'pfl');
+
+    // TODO: Implement loop features
+    // Auto
+    // Shift + Auto
+    // x 1/2
+    // Shift + x 1/2
+    // x2
+    // Shift + x2
+    // Reloop
```

Comment:
I've made a JS object for this for the P32 mapping the other day. https://github.com/Be-ing/mixxx/blob/4dbfa1d5235e54156c2a2b79fccd08410944f694/res/controllers/Hercules-P32-scripts.js#L254